components/sceneutil/shadow.hpp

```diff
@@ -3,19 +3,26 @@
 
 #include <osgShadow/ViewDependentShadowMap>
 
+#include <components/shader/shadermanager.hpp>
+
 namespace SceneUtil
 {
     class MWShadow : public osgShadow::ViewDependentShadowMap
     {
     public:
         static const int numberOfShadowMapsPerLight = 3;
+        static const int enableShadows = true;
         static const bool debugHud = true;
 
         MWShadow();
 
         const static int baseShadowTextureUnit = 8 - numberOfShadowMapsPerLight;
 
         virtual void cull(osgUtil::CullVisitor& cv);
+
+        virtual Shader::ShaderManager::DefineMap getShadowDefines();
```

Comment:
I would have just written out the `std::map<....>` so the include can be removed and the one component doesn't depend on the other. Also, I'm not sure why this function is made virtual?

Or... move the defines setup somewhere else, like RenderingManager. That would be handy if we ever intend to switch back to the upstream version of the shadow technique class.

Reply:
Robert Osfield has pretty much said that osgShadow is abandoned and is going to stay abandoned and anyone wanting to use shadows with OSG should roll their own solution, so I don't think there's much chance at all of the upstream version being better than ours.

Reply:
I don't share your outlook on that. Even if abandoned in Robert's eyes, this is open source where anyone can contribute. And after all, the upstream version worked, too (what we have is mostly that just with a backport of one unreleased fix, right?)

Reply:
The intended method of supporting non-shadow-receiving children isn't implemented in upstream (which I implemented but then broke somehow, so we went with the other option) and if someone fixes that, our code has to not include those changes unless we're going to change all our nodemasks. I've also implemented a working debug HUD that upstream doesn't have, and before this gets merged, there's an issue with projective shadow mapping that needs fixing. By the time this gets merged, it won't be trivially similar to upstream.

Reply:
> The intended method of supporting non-shadow-receiving children isn't implemented in upstream (which I implemented but then broke somehow, so we went with the other option) and if someone fixes that, our code has to not include those changes unless we're going to change all our nodemasks.

We wouldn't have to change anything as long as we keep 'receivesShadowsMask' set to ~0 as it is now.

This discussion is getting kinda out of hand. The upstream-thingy was just a minor potential supporting point for moving the defines to RenderingManager. I could think of other reasons to have all defines set in a central place: we don't have to bother merging different maps, it's easier to find out what the full list of defines is, and the components won't be glued to to a specific way of calling their shader defines (or even that shaders exist at all), which arguably makes them more reusable.

Reply:
My main justification for putting the getter for the shadow defines in the shadow class is that really what those defines are is an implementation detail of the shadow class. It already changed once when I switched from `osgShadow::LightSpacePerspectiveShadowMap` to `osgShadow::ViewDependentShadowMap` as the base class, and it's not impossible that it changes again (although if it does it'll probably be because we end up with a completely custom shadow solution, which isn't what I'm doing now this nearly works). I'm just not convinced that things should end up responsible for shadows if they're not the intuitive place for that kind of thing to go.

As for putting the defines in `RenderingManager` specifically, as it's an OpenMW file, it means if we ever add shadows to the CS, we have to duplicate things. This is something that might be a good idea in the long run as shadows can be a good depth cue when positioning items.